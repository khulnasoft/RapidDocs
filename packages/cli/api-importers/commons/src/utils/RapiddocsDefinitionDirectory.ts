import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { RelativeFilePath, sep } from "@khulnasoft/path-utils";

export class RapiddocsDefinitionDirectory {
    private files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};
    private directories: Record<string, RapiddocsDefinitionDirectory> = {};

    public getAllFiles(): Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> {
        const files: Record<RelativeFilePath, RawSchemas.DefinitionFileSchema> = {};

        const walk = (root: RapiddocsDefinitionDirectory, currentPath?: string) => {
            for (const [relativeFilePath, definition] of Object.entries(root.files)) {
                const fullRelativeFilePath =
                    currentPath != null
                        ? RelativeFilePath.of(`${currentPath}${sep}${relativeFilePath}`)
                        : RelativeFilePath.of(relativeFilePath);
                files[fullRelativeFilePath] = definition;
            }
            const sortedDirectories = Object.keys(root.directories).sort();
            for (const directory of sortedDirectories) {
                const nextPath = currentPath != null ? `${currentPath}${sep}${directory}` : directory;
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const nextDirectory = root.directories[directory]!;
                walk(nextDirectory, nextPath);
            }
        };

        walk(this);

        return files;
    }

    public getOrCreateFile(relativeFilePath: RelativeFilePath): RawSchemas.DefinitionFileSchema {
        return this.getOrCreateFileRecursive(relativeFilePath.split(sep));
    }

    private getOrCreateFileRecursive(pathParts: string[]): RawSchemas.DefinitionFileSchema {
        if (pathParts.length === 1) {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            return (this.files[RelativeFilePath.of(pathParts[0]!)] ??= {});
        }
        const [directory, ...remainingPath] = pathParts;
        if (directory == null) {
            throw new Error(`Internal error; cannot add file with path: ${pathParts}`);
        }
        if (!this.directories[directory]) {
            this.directories[directory] = new RapiddocsDefinitionDirectory();
        }
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return this.directories[directory]!.getOrCreateFileRecursive(remainingPath);
    }
}
