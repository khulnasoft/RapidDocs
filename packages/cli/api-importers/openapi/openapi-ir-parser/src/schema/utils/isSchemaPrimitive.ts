import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaPrimitive(schema: Schema): boolean {
    if (schema.type === "primitive") {
        return true;
    } else if (schema.type === "nullable" || schema.type === "optional") {
        return isSchemaPrimitive(schema.value);
    }
    return false;
}
