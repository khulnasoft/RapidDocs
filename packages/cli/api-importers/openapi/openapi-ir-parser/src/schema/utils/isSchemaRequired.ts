import { Schema } from "@khulnasoft/openapi-ir";

export function isSchemaRequired(schema: Schema): boolean {
    return schema.type !== "optional" && schema.type !== "nullable";
}
