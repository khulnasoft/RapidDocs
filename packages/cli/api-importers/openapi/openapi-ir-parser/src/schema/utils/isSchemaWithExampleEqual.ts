import { SchemaWithExample, isSchemaEqual } from "@khulnasoft/openapi-ir";

import { convertSchemaWithExampleToSchema } from "./convertSchemaWithExampleToSchema";

// only diffs the schema, not the example
export function isSchemaWithExampleEqual(a: SchemaWithExample, b: SchemaWithExample): boolean {
    return isSchemaEqual(convertSchemaWithExampleToSchema(a), convertSchemaWithExampleToSchema(b));
}
