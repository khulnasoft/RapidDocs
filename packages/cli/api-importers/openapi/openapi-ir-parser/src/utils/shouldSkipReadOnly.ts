import { RapiddocsOpenapiIr } from "@khulnasoft/openapi-ir";

export function shouldSkipReadOnly(method: RapiddocsOpenapiIr.HttpMethod): boolean | undefined {
    return method === "POST" || method === "PUT" || method === "PATCH";
}
