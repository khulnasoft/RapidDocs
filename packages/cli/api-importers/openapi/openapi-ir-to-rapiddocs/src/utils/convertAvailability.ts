import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Availability } from "@khulnasoft/openapi-ir";

export function convertAvailability(
    availability: Availability | undefined
): RawSchemas.AvailabilityUnionSchema | undefined {
    switch (availability) {
        case Availability.Deprecated:
            return "deprecated";
        case Availability.Beta:
            return "pre-release";
        case Availability.GenerallyAvailable:
            return "generally-available";
        default:
            return undefined;
    }
}
