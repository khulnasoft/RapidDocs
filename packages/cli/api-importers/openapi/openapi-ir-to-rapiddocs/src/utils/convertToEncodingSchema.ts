import { assertNever } from "@khulnasoft/core-utils";
import { RawSchemas } from "@khulnasoft/rapiddocs-definition-schema";
import { Encoding } from "@khulnasoft/openapi-ir";

export function convertToEncodingSchema(encoding: Encoding): RawSchemas.EncodingSchema {
    switch (encoding.type) {
        case "protobuf":
            return {
                proto: {
                    type: encoding.typeName
                }
            };
        default:
            assertNever(encoding.type);
    }
}
