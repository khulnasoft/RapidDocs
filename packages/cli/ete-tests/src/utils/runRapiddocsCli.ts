import { Options } from "execa";
import path from "path";

import { loggingExeca } from "@khulnasoft/logging-execa";

export async function runRapiddocsCli(args: string[], options?: Options): Promise<loggingExeca.ReturnValue> {
    return loggingExeca(undefined, "node", [path.join(__dirname, "../../../cli/dist/dev/cli.cjs"), ...args], {
        ...options,
        env: {
            ...options?.env,
            RAPIDDOCS_TOKEN: process.env.RAPIDDOCS_ORG_TOKEN_DEV
        },
        doNotPipeOutput: options?.reject === false
    });
}
