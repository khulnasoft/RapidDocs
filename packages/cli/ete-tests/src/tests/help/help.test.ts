import tmp from "tmp-promise";

import { runRapiddocsCli } from "../../utils/runRapiddocsCli";

describe("help", () => {
    it("no arguments", async () => {
        const { stderr, failed } = await runRapiddocsCli([], {
            cwd: (await tmp.dir()).path,
            reject: false
        });
        expect(stderr).toMatchSnapshot();
        expect(failed).toBe(true);
    }, 60_000);
});
