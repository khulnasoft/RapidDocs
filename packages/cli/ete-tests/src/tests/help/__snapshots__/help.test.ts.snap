// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`help > no arguments 1`] = `
"rapiddocs-dev

Commands:
  rapiddocs-dev init                         Initialize a Rapiddocs API
  rapiddocs-dev token                        Generate a Rapiddocs Token
  rapiddocs-dev add <generator>              Add a code generator to generators.yml
  rapiddocs-dev generate                     Generate all generators in the specified
                                        group
  rapiddocs-dev ir <path-to-output>          Generate IR (Intermediate
                                        Representation)
  rapiddocs-dev check                        Validates your Rapiddocs Definition. Logs
                                        errors.
  rapiddocs-dev login                        Log in to Rapiddocs via GitHub
  rapiddocs-dev format                       Formats your Rapiddocs Definition
  rapiddocs-dev write-definition             Write underlying Rapiddocs Definition for
                                        OpenAPI specs and API Dependencies.
  rapiddocs-dev docs                         Commands for managing your docs
  rapiddocs-dev mock                         Starts a mock server for an API.
  rapiddocs-dev write-overrides              Generate a basic openapi overrides file.
  rapiddocs-dev test                         Runs tests specified in --command, this
                                        spins up a mock server in the background
                                        that is terminated upon completion of
                                        the tests.
  rapiddocs-dev api update                   Pulls the latest OpenAPI spec from the
                                        specified origin in generators.yml and
                                        updates the local spec.
  rapiddocs-dev upgrade                      Upgrades version in rapiddocs.config.json.
                                        Also upgrades generators in
                                        generators.yml to their
                                        minimum-compatible versions.
  rapiddocs-dev jsonschema <path-to-output>  Generate JSON Schema for a specific type
  rapiddocs-dev generator                    Operate on the generators within your
                                        Rapiddocs configuration

Options:
      --help       Show help                                           [boolean]
      --log-level
          [choices: "trace", "debug", "info", "warn", "error"] [default: "info"]
  -v, --version    Print current version"
`;
